### ISSUE TITLE
cloud/https: use LetsEncrypt http-01

### ISSUE THREAD

**Issue by adg** [upspin/upspin#568]
A recently found security vulnerability in the Let's Encrypt protocol means that we can no longer use the tls-sni-01 challenge to issue TSL certs. This means we need to use http-01, which means all Upspin servers that use Let's Encrypt certs need to serve port 80 as well as 443. This will require several changes to our code and the infrastructure of everyone who has deployed Upspin with Let's Encypt certs. This issue tracks those changes.

See golang/go#21890 for discussion, and the change to the x/crypto/acme/autocert package. And also [the announcement](https://community.letsencrypt.org/t/2018-01-11-update-regarding-acme-tls-sni-and-shared-hosting-infrastructure/50188) about tls-sni-01 being disabled forever.

Pending changes, list updated as I think of and/or fix them:

- [x] Package upspin.io/cloud/https must serve on both port 80 and 443, with options to configure the ports for forwarding configurations.
- [x] Documentation for server deployment must be updated to instruct users to forward port 80 as well as port 443.
- [x] The Dockerfiles in gcp.upspin.io/cloud/docker must be updated to expose port 80.
- [x] The Kubernetes config files in gcp.upspin.io/cloud/kubernetes must be updated to expose/forward port 80.
- [x] Re-deploy the upspin.io dir, store, and key servers.
- [x] The upspin-ui server deployment tool must create a firewall rule to allow port 80 and also adjust its cloud-config script 
- [x] Provide instructions for people who have deployed with upspin-ui (they need to adjust firewall rules and the cloud-config script for their VM; maybe we can provide a script to automate this).
- [x] Provide instructions for people who have deployed upspinservers manually (in essence, tell them to allow connections to port 80, or forward connections to port 80 if they're behind a firewall).

**Comment by adg** [comment]
For people that deployed an upspinserver with upspin-ui, they can use this script to make the relevant changes: https://gist.github.com/adg/304f7ce803981856eca608c05d3b190e

**Comment by adg** [comment]
Sent mail: https://groups.google.com/forum/#!msg/upspin/NrkLzlA1LQs/CmeoJsGuAQAJ

**Comment by filmil** [comment]
Any instructions on how to verify that the changes have been applied correctly?  E.g. is there an example `curl` command line that I can use to verify that upspin is now serving proper responses?

**Comment by adg** [comment]
@filmil if you curl your server with an `http://` scheme on port 80, it should return a redirect to port HTTPS. If that works, you're set.

**Comment by filmil** [comment]
Is there a way to express that in terms of the `curl` command output, so that it can be tested automatically?

For example, if I do a `curl http://my.server.addr.tld` I get a 302 to an LAN hostname, and port 443.  Is this enough?  What should this name be instead if the answer is "no"?  What if I run `upspinserver` where both directory and store server are same machine?  Does the redirect need to work on both?  What should the redirect targets be?

I ask because I have a limited supply of IP addresses (one), and have to run a reverse proxy.  So I need to get the setup exactly right for things to continue working, and it's much easier to do if I know what the expected result should be, versus trial and error.  I get that a GCP-hosted setup works fine with a minor intervention, but this isn't it.

**Comment by adg** [comment]
Your `upspinserver` should have a host name, specified by the `Addr` field in `$HOME/upspin/server/serverconfig.json` on your server. Public HTTP and HTTPS requests to that host name should be proxied to the `upspinserver`.

When `upspinserver` receives an HTTP request it will redirect it to HTTPS using the host name specified in `serverconfig.json`. So that name must not be a name known only to your LAN. 

Here's how it looks for my Upspin server, `upspin.adg.tech`:

```
$ curl -vv http://upspin.adg.tech
* Rebuilt URL to: http://upspin.adg.tech/
*   Trying 43.229.60.243...
* TCP_NODELAY set
* Connected to upspin.adg.tech (43.229.60.243) port 80 (#0)
> GET / HTTP/1.1
> Host: upspin.adg.tech
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 302 Found
< Content-Type: text/html; charset=utf-8
< Location: https://upspin.adg.tech/
< Date: Sun, 21 Jan 2018 21:56:59 GMT
< Content-Length: 47
< 
<a href="https://upspin.adg.tech/">Found</a>.
```

**Comment by filmil** [comment]
@adg Thanks, this is exactly the test that I needed!  And could confirm that now my upspin servers are set up correctly.

Some back story, I think I may have made a mistake initially when I created two separate server FQDNs for the key server and storage server even though it's one physical machine.  So now, I needed a way to plumb both of them through the reverse proxy on my machine.  No big deal in retrospect, but I did spend a few minutes scratching my head on what to do.

This test made it possible to test a few approaches quickly.  Thanks again!
